File: 46-permutations/46-permutations.go
func permute(nums []int) [][]int {
    n := len(nums)
    var (
        dfs func(int)
        res [][]int
    )
    dfs = func(idx int) {
        if idx == n {
            res = append(res, append([]int{}, nums...))
            return
        }
        for i := idx; i < n; i++ {
            nums[idx], nums[i] = nums[i], nums[idx]
            dfs(idx+1)
            nums[idx], nums[i] = nums[i], nums[idx]
        }
    }
    dfs(0)
    return res
}

File: 46-permutations/46-permutations.py
# time: O(n!)
# space: O(n!)

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        def helper(idx: int):
            nonlocal res
            nonlocal n
            if idx == n:
                res.append(nums[:])
                return
            for i in range(idx, n):
                nums[idx], nums[i] = nums[i], nums[idx]
                helper(idx+1)
                nums[idx], nums[i] = nums[i], nums[idx]
            
        helper(0)
        return res
